% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/circIMPACT_functions.R
\name{marker.selection}
\alias{marker.selection}
\title{Select circRNA markers
To establish a possible impact of circRNA in gene expression
we select a subset of circRNA defined as markers: 
only circRNA which expression can discretize two groups 
(DESeq adj. p-value<.01) have been selected.}
\usage{
marker.selection(
  dat,
  dds,
  sf,
  p.cutoff = 0.01,
  lfc.cutoff = NULL,
  method.d,
  method.c,
  k,
  choose.k = FALSE,
  index.m = NULL,
  max.nc = 2,
  plot = FALSE,
  n = 9,
  median = TRUE
)
}
\arguments{
\item{dat}{circRNA data set}

\item{dds}{DeseqDataSet object filterd}

\item{sf}{sizefactor}

\item{p.cutoff}{select only circRNAs signficantly differentially expressed}

\item{lfc.cutoff}{log fold change cutoff to select circRNAs highly deregulated}

\item{method.d}{method for calculate matrix distance}

\item{method.c}{method for clustering}

\item{k}{number of group in K-means algorithm}
}
\value{
list of two elements: circRNAs estimate, circRNA-target ids
}
\description{
Select circRNA markers
To establish a possible impact of circRNA in gene expression
we select a subset of circRNA defined as markers: 
only circRNA which expression can discretize two groups 
(DESeq adj. p-value<.01) have been selected.
}
