---
title: "The CircTarget package - exploring molecular pathways by linking circular RNA to TARGET genes"
author: 
  - name: Alessia Buratin
    affiliation: Departement of Biology, Department of Molecular Medicine, University of Padova
date: "`r format(Sys.Date(), '%d-%m-%Y')`"    
output:
  prettydoc::html_pretty:
    theme: hpstr  
    highlight: github
bibliography: library.bib
link-citations: yes
vignette: >
  %\VignetteIndexEntry{CircTarget}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

<style>
    body {
        max-width: 1500px;
        text-align: justify
    }
    .page-header {
      background-image: none;
      background-color: #ff7600;
    }
    .main-content h2 {
      color: #ff7600;
    }
</style>


## Overview

*CircTarget* provides a strategy to analyze circular and linear transcriptome rely on data mining and statistical learning techniques. *CircTarget* package offers a tidy pipeline, starting from the provided circRNA-Target which expression classifies samples with high and low expression of it, a comprensive transcriptome and molecular pathway analysis are performed, including visualisation, normalization, classification, clustering, Differential Expression Analysis and Gene Set Enrichment Analysis. The package accepts data presented as a matrix of raw counts and allows the inclusion of variables that occur with the experimental setting. A series of functions enables data cleaning by filtering rows, data adjustment by identify and removing the unwonted source of variation and to select the best predictors for modeling the responce variable. A learning technique is applied to build a robust classification model, and also an unsupervised analysis is carried out to gain mechanistic insights and detect the molecular pathway associated with the expression levels of the Target circRNA. Finally, a Differential Expression Analysis identyfied deregulated mRNAs and a GSEA is performed for them. *CircTarget* stands for "circRNA-Target".
CircTarget package version: `r packageVersion("CircTarget")`

*CircTarget* provides an R interface to analyze a possible impact of circRNA expression profile in gene expression.

The toolset performs functional enrichment analysis and visualization of gene lists obtained comparing samples with low circRNAs expression with those with high expression. 

The main tools in *CircTarget* are:

* `marker.detection` - detection of circRNAs that can stratified samples by their expression signifincantly
* `geneexpression` - DEG analysis for each circRNA-markers defined by K-means algorithm or by circRNA expression (high vs low)

The input for any of the tools consist in count matrices: from the same samples we need the quantification of circular and linear RNA.

----

## Installation and loading

```{r eval = FALSE}
devtools::install_github("ABuratin/CircTarget")
```

```{r setup, warning = FALSE, message = FALSE}
library(CircTarget)
library(DESeq2)
library(data.table)
library(plyr)
library(Rtsne)
library(ggplot2)
library(ggrepel)
library(plotly)
library(ComplexHeatmap)
library(circlize)
library(viridis)
library(knitr)
library(kableExtra)
library(formattable)
library(htmltools)
library(sparkline)
library(tidyverse)
library(RColorBrewer)
library(purrr)
library(magrittr)
library(webshot)
```
----

### Input data

```{r}
data("circularData")
data("count.matrix")
data("meta")
data("coldata.df")
## define sample/condition colors
intgroup.dt <- meta[, .(.N), by = .(sample, 
                                      condition)][order(sample), 
                                                  .(sample), by = condition]
samples.per.condition <- data.frame(intgroup.dt[, .N, by = .(condition)], row.names = "condition")
  
n.conditions <- nrow(samples.per.condition)
hues <- CircTarget::gg_color_hue(n.conditions)
for(i in 1:n.conditions){
      n.hues <- samples.per.condition[i, "N"] + 2
      col.hues <- colorRampPalette(colors = c(hues[i], 
                                              "white"))(n.hues)[1:(n.hues-2)]
      
      intgroup.dt[condition == rownames(samples.per.condition)[i], `:=`(color = col.hues,
                                                                        hue = hues[i])]
    }
## create a deseq dataset object 
dds.circular <- suppressMessages(DESeqDataSetFromMatrix(countData = ceiling(circularData[, coldata.df$sample[order(coldata.df$condition)]]),
                                   colData = coldata.df[order(coldata.df$condition),],
                                   design = ~ condition))
dds.circular <- suppressMessages(estimateSizeFactors(dds.circular))
sf <- sizeFactors(dds.circular)
```

At first a simply filter of low count is applied to remove background noise.

```{r}
data.filt <- circularData[rowSums(circularData >= 5) >= 2,]
dds.filt.expr <- suppressMessages(DESeqDataSetFromMatrix(countData = ceiling(data.filt[,coldata.df$sample[order(coldata.df$condition)]]),
                                   colData = coldata.df[order(coldata.df$condition),],
                                   design = ~ condition))
dds.filt.expr <- suppressMessages(estimateSizeFactors(dds.filt.expr))
sf.filt <- sizeFactors(dds.filt.expr)
circNormDeseq <- counts(dds.filt.expr, normalized = T)
```

## Selection of circRNA-markers with `marker.detection`

### CircRNA-markers

To establish a possible impact of circRNA in gene expression using different molecular subtypes of human T-ALL we select a subset of circRNA defined as markers: at first by using k-means algorithm we separate samples in two main cluster defined by the circRNA expression, then only circRNA which DESeq adj. p-value<.1 and log fold-change>1.5 have been selected.

```{r message=FALSE, warning=FALSE, include=TRUE}

circtarget <- marker.selection(dat = circNormDeseq, dds = dds.filt.expr, sf = sf.filt, p.cutoff = 0.1, lfc.cutoff = 1, 
                                 method.d = "spearman", method.c = "complete", k = 2, median = TRUE)
```

The result `list` will include:

* circ.markers - a data frame with DESeq estimates for each circRNAs.  
* circ.targets - a vector of circRNAs defined as possible targets.
* group.df - data.frame with a variable group that indicate the new stratification of samples by circRNAs expression.
* plot - formatted table with info. about circTarget expression

### Density expression of circRNA-markers

```{r}
circMark <- circtarget$circ.targetIDS[5]
circMark_group.df <- circtarget$group.df[circtarget$group.df$circ_id==circMark,]
circMark_group.df$counts <- merge(circMark_group.df, reshape2::melt(circNormDeseq[circMark,]), by.x = "sample_id", by.y = "row.names")[,"value"]
mu <- ddply(circMark_group.df, "group", summarise, Mean=mean(counts), Median=median(counts), Variance=var(counts))

p <- ggplot(circMark_group.df, aes(x=counts, color=group, fill=group)) +
  geom_density(alpha=0.3) + 
  geom_vline(data=mu, aes(xintercept=Median, color=group),
             linetype="dashed") +
  geom_text(data=mu, aes(x=Median[group=="g1"] - 0.55, 
                         label=paste0("Median:", round(Median[group=="g1"], 3), " Variance:", round(Variance[group=="g1"]), 3), y=0.15),
            colour="black", angle=90, text=element_text(size=9)) +
  geom_text(data=mu, aes(x=Median[group=="g2"] - 0.55, 
                       label=paste0("Median:", round(Median[group=="g2"], 3), " Variance:", round(Variance[group=="g2"]), 3 ), y=0.15), 
          colour="black", angle=90, text=element_text(size=11)) +  scale_fill_brewer(palette="Dark2") + 
  scale_color_brewer(palette="Dark2") + 
  labs(title=paste0("circMarker (", circMark, ")", " counts density curve"), x = "Normalized read counts", y = "Density") + 
  theme_classic()
p

```


### Clustering analysis using CircRNA-markers

Clustering analysis is performed using the previuos selected circRNA, defined as circRNA-markers. 
T-SNE algorithm is used to performe dimensionality reduction and k-means and hierarchical clustering methods are compared in order to identify two cluster of samples. 

```{r message=FALSE}
markers.circrnas <- circtarget$circ.targetIDS
mat.filt.mark <- circularData[markers.circrnas, ]

dds.filt.mark <- DESeqDataSetFromMatrix(countData = ceiling(mat.filt.mark[,coldata.df$sample]),
                                   colData = coldata.df,
                                   design = ~ 1)
dds.filt.vst <- varianceStabilizingTransformation(dds.filt.mark, fitType = "local", blind = F)
norm.counts.filt <- assay(dds.filt.vst)
```

```{r }
dt <- norm.counts.filt

```

```{r}
## Rtsne function may take some minutes to complete...
set.seed(9)
mydist <- dist(t(norm.counts.filt))
## t-SNE representation
# set a perplexity parameter consistent with the number of samples
tsne_data <- Rtsne(mydist, pca = F, perplexity=1, max_iter=5000)

## getting the two dimension matrix
d_tsne_1 = as.data.frame(tsne_data$Y)
rownames(d_tsne_1) <- colnames(norm.counts.filt)

```

```{r clustering, fig.cap = "t-SNE dimensionality reduction representation. K-means and hierarchical clustering are compared."}

## keeping original data
d_tsne_1_original=d_tsne_1

## Creating k-means clustering model, and assigning the result to the data used to create the tsne
fit_cluster_kmeans=kmeans(scale(d_tsne_1), 2)
d_tsne_1_original$cl_kmeans = factor(fit_cluster_kmeans$cluster)

## Creating hierarchical cluster model, and assigning the result to the data used to create the tsne
fit_cluster_hierarchical=hclust(dist(scale(d_tsne_1)))

## setting 2 clusters as output
d_tsne_1_original$cl_hierarchical = factor(cutree(fit_cluster_hierarchical, k=2))

# Plotting the cluster models onto t-SNE output

plot_cluster=function(data, var_cluster, palette)
{
  ggplot(data, aes_string(x="V1", y="V2", color=var_cluster)) +
  geom_point(size=3) +
  guides(colour=guide_legend(override.aes=list(size=3))) +
  geom_text_repel(aes(label = rownames(data)), 
                  hjust = 0.5, vjust = -1) +
  xlab("") + ylab("") +
  ggtitle("") +
  theme_light(base_size=11) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank(),
        legend.direction = "horizontal", 
        legend.position = "bottom",
        legend.box = "horizontal") + 
    scale_colour_brewer(palette = palette) 
}


plot_k=plot_cluster(d_tsne_1_original, "cl_kmeans", "Dark2")
plot_h=plot_cluster(d_tsne_1_original, "cl_hierarchical", "Set1")

## and finally: putting the plots side by side with gridExtra lib...
library(gridExtra)
grid.arrange(plot_k, plot_h,  ncol=2)

```


```{r}
pca <- prcomp(x = t(norm.counts.filt), center = T)
d <- data.frame(pca$x[rownames(coldata.df), c("PC1", "PC2")], coldata.df)
PC1.var <- summary(pca)$importance["Proportion of Variance", 1]
PC2.var <- summary(pca)$importance["Proportion of Variance", 2]
g1 <- ggplot(data = d, 
       mapping = aes(x = PC1, y = PC2)) +
    geom_point(size = 4) +
    coord_fixed(ratio = 1) +
    xlab(paste0("PC1: ", percent(PC1.var))) +
    ylab(paste0("PC2: ", percent(PC2.var))) +
    theme_classic() + 
    theme(legend.position = "bottom", 
          plot.title = element_text(hjust = .5))
library(factoextra)
#### compute contribution 
contrib <- function(ind.coord, comp.sdev, n.ind){
  100*(1/n.ind)*ind.coord^2/comp.sdev^2
}
ind.contrib <- t(apply(pca$x, 1, contrib, 
                       pca$sdev, nrow(pca$x)))
g3 <- fviz_contrib(pca, choice="var", axes = 1:3, top = 8)
library(ggpubr)

ggpubr::ggarrange(ggarrange(g1, g3, 
                            ncol = 2, 
                            labels = c("A", "B")),
                  nrow = 1)  
```

### Heatmap using CircRNA-markers

```{r }

set.seed(201)
dds.filt.mark <- estimateSizeFactors(dds.filt.mark)
circNormDeseq <- counts(dds.filt.mark, normalized = T)

base_mean = log2(rowMeans(circNormDeseq)+0.001)
mat_scaled = t(apply(dt, 1, function(x) scale(x = x, center = T, scale = T)))
colnames(mat_scaled) <- colnames(dt)
cond = colData(dds.filt.expr)$condition
## choice of kmeans results as cluster of samples
clus = d_tsne_1_original$cl_kmeans
cond.colors <- unique(intgroup.dt$hue)
names(cond.colors) <- unique(intgroup.dt$condition)
ha = HeatmapAnnotation(df = data.frame(condition = cond, cluster = clus),
                       col = list(condition = cond.colors),
                       show_annotation_name = F,
                       annotation_legend_param = list(condition = list(nrow = 2, direction = "horizontal")))

mat.dend <- as.dendrogram(fit_cluster_hierarchical)
fit_cluster_kmeans$cluster  
ht <- Heatmap(mat_scaled, name = "expression", 
        # km = 2,
        # column_km = 2,
        column_order = names(fit_cluster_kmeans$cluster[order(fit_cluster_kmeans$cluster)]),
        col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
        top_annotation = ha, 
        # top_annotation_height = unit(4, "mm"),
        clustering_distance_columns = "euclidean",
        clustering_method_column = "complete",
        cluster_columns = F,
        clustering_distance_rows = "spearman",#"minkowski",
        clustering_method_rows = "ward.D2",
        cluster_rows = T,
        # row_dend_side = "right",
        # row_names_side = "left",
        show_row_names = T, 
        show_column_names = F, 
        width = unit(9, "cm"),
        show_row_dend = T,
        show_column_dend = T,
        # row_dend_reorder = TRUE,
        row_names_gp = gpar(fontsize = 5),
        heatmap_legend_param = list(direction = "horizontal")) +
Heatmap(base_mean, name = "log2(base mean)", show_row_names = F, width = unit(2, "mm"), col = inferno(255), show_column_names = F, row_names_gp = gpar(fontsize = 5), heatmap_legend_param = list(direction = "horizontal"))

draw(ht, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

```


## Gene list functional enrichment analysis with *CircTargets*

### Gene expression analysis - Table results

Because of these analysis is highly time consuming for this tutorial have been reported the DEGs for only the first 3 circRNA-markers.

```{r Filterlinear_data}
## filter out genes low expressed 
min.count <- 20
min.col <- 5

filt.mat <- count.matrix[rowSums(count.matrix >= min.count) >= min.col, ]

```

```{r message=FALSE, warning=FALSE}
#normalized circRNAs data 
circNormDeseq <- counts(dds.filt.expr, normalized = T) %>% as.data.frame()
circNormDeseq$circ_id <- rownames(circNormDeseq)

library(doParallel)
no_cores <- detectCores() - 1  
registerDoParallel(cores=no_cores)  

gene_mark <- foreach::foreach(i=1:5, .combine = rbind) %dopar% {

  results.temp <- data.frame(geneexpression(circ_idofinterest = markers.circrnas[i], circRNAs = circNormDeseq, 
                                       linearRNAs = filt.mat, colData = coldata.df, padj = 0.1, 
                                       group = circtarget$group.df[circtarget$group.df$circ_id%in%markers.circrnas[i],],
                                       covariates = NULL), circTarget = markers.circrnas[i])
}

gene_mark_hipk3 <- data.frame(geneexpression(circ_idofinterest = "11:33286412-33287511", circRNAs = circNormDeseq, 
                                       linearRNAs = filt.mat, colData = coldata.df, padj = 0.1, 
                                       group = circtarget$group.df[circtarget$group.df$circ_id%in%"11:33286412-33287511",],
                                       covariates = NULL), circTarget = "11:33286412-33287511")

```


```{r}

gene_mark <- as.data.table(gene_mark_hipk3)
gene_mark %>% dplyr::rename("Gene" = "gene_id", "logFC" = "log2FoldChange") %>% 
  arrange(padj) %>% 
  select(circTarget, Gene, logFC) %>% head(20) %>% 
  formattable::formattable(., align = c("c","c","c"), list(
          gene_id = formattable::formatter("span", style = ~ formattable::style(color = "grey", font.weight = "bold")),
          circTarget = formattable::formatter("span", style = ~ formattable::style(color = "grey", font.weight = "bold")),
          logFC = CircTarget::color_tile3(digits = 3, n = 18, fun = "comma", palette = "PiYG")))

# knitr::kable(gene_mark %>% dplyr::group_by(circRNA_markers, n.degs) %>% 
# dplyr::summarise(DEGs = paste(sort(gene_id),collapse=", ")),
#       escape = F, align = "c", row.names = T, caption = "circRNA-DEGs assosiation") %>% kable_styling(c("striped"), full_width = T)

```

### Enrichment analysis

```{r}
#subset gene symbol deregulated using the interesting circRNA marker as stratificator
geneList <- gene_mark$log2FoldChange[gene_mark$circTarget==markers.circrnas[5]]
geneList <- gene_mark_hipk3$log2FoldChange

# order gene list by foldchange
geneList = sort(geneList, decreasing = TRUE)
names(geneList) <- gene_mark$Gene[gene_mark$circTarget==markers.circrnas[5]]
names(geneList) <- gene_mark_hipk3$gene_id

library(gprofiler2)

gostres2 <- gost(query = names(geneList)[names(geneList)!="."], 
                 organism = "hsapiens", ordered_query = TRUE, 
                 multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                 measure_underrepresentation = FALSE, evcodes = TRUE, 
                 user_threshold = 0.05, correction_method = "g_SCS", 
                 domain_scope = "annotated", custom_bg = NULL, 
                 numeric_ns = "", sources = NULL)
p <- gostplot(gostres2, capped = FALSE, interactive = TRUE)
p
```

